chore(docs): MDX v2 (#35893)
gatsbyjs/gatsby: packages/gatsby-plugin-mdx/README.md

# gatsby-plugin-mdx

`gatsby-plugin-mdx` is the official integration for using [MDX](https://mdxjs.com) with [Gatsby](https://www.gatsbyjs.com).

MDX is markdown for the component era. It lets you write JSX embedded inside markdown. It’s a great combination because it allows you to use markdown’s often terse syntax (such as `# heading`) for the little things and JSX for more advanced components.

## Table of contents

- [Installation](#installation)
- [Usage](#usage)
  - [Configuration](#configuration)
    - [Extensions](#extensions)

>     - [Gatsby remark plugins](#gatsby-remark-plugins)

Reviewer: ```suggestion
    - [Gatsby remark plugins](#gatsby-remark--plugins)
```

Fixes link

    - [mdxOptions](#mdxoptions)
  - [Imports](#imports)
  - [Layouts](#layouts)
  - [Programmatically create MDX pages](#programmatically-create-mdx-pages)
  - [GraphQL MDX Node structure](#graphql-mdx-node-structure)
  - [Extending the GraphQL Mdx nodes](#extending-the-graphql-mdx-nodes)
  - [Components](#components)
    - [MDXProvider](#mdxprovider)
    - [Shortcodes](#shortcodes)
- [Migrating from v3 to v4](#migrating-from-v3-to-v4)
- [Why MDX?](#why-mdx)
- [Related](#related)

## Installation

```shell
npm install gatsby-plugin-mdx @mdx-js/react gatsby-source-filesystem
```

## Usage

After installing `gatsby-plugin-mdx` you can add it to your plugins list in your

> `gatsby-config.js`. You'll also want to configure `gatsby-source-filesytem` to point at your `src/pages` directory (even if you don't want to create MDX pages from `src/pages`).

Reviewer: Does the order of the plugins matter? Noticed the order changed in the diff

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-plugin-mdx`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `pages`,
        path: `${__dirname}/src/pages`,
      },
    },
  ],
}
```

By default, this configuration will allow you to automatically create pages with `.mdx` files in `src/pages` and will process any Gatsby nodes with Markdown media types into MDX content.

Note that `gatsby-plugin-mdx` requires `gatsby-source-filesystem` to be present and configured to process local markdown files in order to generate the resulting Gatsby nodes.

> To automatically create pages with `.mdx` from other sources, you also need to configure `gatsby-plugin-page-creator`.

Reviewer: Would they need to use this instead of an additional `gatsby-source-filesystem`? It doesn't look like there's mention of `gatsby-plugin-page-creator` in the dotcom docs page

Reviewer 2: If you want to have MDX files automatically turned into pages outside of `src/pages` (which is a bit of an uncommon use case) you'll need to configure the `gatsby-source-filesystem` and `gatsby-plugin-page-creator`.

I think "from other sources" is misleading here, it's only relevant for other local sources in another location

Reviewer 2: ```suggestion
To automatically create pages from MDX files outside of `src/pages` you'll need to configure `gatsby-plugin-page-creator` and `gatsby-source-filesystem` to point to this folder of files.
```

Reviewer 2: Better like that?

Reviewer: Yeah that's clearer to me now, thanks!

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    `gatsby-plugin-mdx`,
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `posts`,
        path: `${__dirname}/src/posts`,
      },
    },
    {
      resolve: `gatsby-plugin-page-creator`,
      options: {
        path: `${__dirname}/src/posts`,
      },
    },
  ],
}
```

Also check out the guide [Adding MDX Pages](https://www.gatsbyjs.com/docs/how-to/routing/mdx/) for more details.

### Configuration

`gatsby-plugin-mdx` exposes a configuration API that can be used similarly to any other Gatsby plugin. You can define MDX extensions, layouts, global scope, and more.

| Key                                             | Default    | Description                                                         |
| ----------------------------------------------- | ---------- | ------------------------------------------------------------------- |
| [`extensions`](#extensions)                     | `[".mdx"]` | Configure the file extensions that `gatsby-plugin-mdx` will process |

> | [`gatsbyRemarkPlugins`](#gatsby-remark-plugins) | `[]`       | Use Gatsby-specific remark plugins                                  |

Reviewer: ```suggestion
| [`gatsbyRemarkPlugins`](#gatsby-remark--plugins) | `[]`       | Use Gatsby-specific remark plugins                                  |
```

Fixes link

}
```

#### `gatsby-remark-*` plugins

> This config option is used for compatibility with a set of plugins many people [use with remark](https://www.gatsbyjs.com/plugins/?=gatsby-remark-) that require the gatsby environment to function properly. In some cases, like [gatsby-remark-prismjs](https://www.gatsbyjs.com/plugins/gatsby-remark-prismjs/), it makes more sense to use a library like [prism-react-renderer](https://github.com/FormidableLabs/prism-react-renderer) to render codeblocks using a [React component](/api-reference/mdx-provider). In other cases, like [gatsby-remark-images](https://www.gatsbyjs.com/plugins/gatsby-remark-images/), the interaction with the Gatsby APIs is well deserved because the images can be optimized by Gatsby and you should continue using it.

Reviewer: ```suggestion
This config option is used for compatibility with a set of plugins many people [use with remark](https://www.gatsbyjs.com/plugins/?=gatsby-remark-) that require the Gatsby environment to function properly. In some cases, like [gatsby-remark-prismjs](https://www.gatsbyjs.com/plugins/gatsby-remark-prismjs/), it makes more sense to use a library like [prism-react-renderer](https://github.com/FormidableLabs/prism-react-renderer) to render codeblocks using a [React component](/api-reference/mdx-provider). In other cases, like [gatsby-remark-images](https://www.gatsbyjs.com/plugins/gatsby-remark-images/), the interaction with the Gatsby APIs is well deserved because the images can be optimized by Gatsby and you should continue using it.
```

    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        mdxOptions: {
          remarkPlugins: [
            require("remark-abbr"),
            // To pass options, use a 2-element array with the
            // configuration in an object in the second element
            [require("remark-external-links"), { target: false }],
          ],
          rehypePlugins: [
            // Generate heading ids for rehype-autolink-headings
            require("rehype-slug"),
            // To pass options, use a 2-element array with the
            // configuration in an object in the second element
            [require("rehype-autolink-headings"), { behavior: "wrap" }],
          ],
        },
      },
    },
  ],
}
```

### Imports

> When importing a react component into your MDX, you can import it using the `import`
> statement as in JavaScript.

Reviewer 2: ```suggestion
When importing a React component into your MDX, you can import it using the `import` statement like in JavaScript.
```
